src/main/java/picard/sam/SamToFastqWithTags.java

```diff
@@ -0,0 +1,225 @@
+    public List<String> TAG_GROUP_SEPERATOR;
+
+    @Argument(shortName = "GZOPTG", doc = "Compress output FASTQ files per Tag grouping using gzip and append a .gz extension to the file names.")
+    public Boolean COMPRESS_OUTPUTS_PER_TAG_GROUP = false;
+
+    private final Log log = Log.getInstance(SamToFastqWithTags.class);
+
+    private static final String TAG_SPLIT_DEFAULT_SEP = "";
+    private static final String TAG_SPLIT_QUAL = "~";
+
+    private ArrayList<String[]> SPLIT_SEQUENCE_TAGS;
+    private ArrayList<String[]> SPLIT_QUALITY_TAGS;
+    private ArrayList<String> SPLIT_SEPARATOR_TAGS;
+
+    @Override
+    protected void initializeAdditionalWriters() {
+        setupTagSplitValues();
+    }
+
+    @Override
+    protected void handleAdditionalRecords(SAMRecord currentRecord, Map<SAMReadGroupRecord, List<FastqWriter>> tagWriters, SAMRecord read1, SAMRecord read2) {
+        final List<FastqWriter> rgTagWriters = tagWriters.get(currentRecord.getReadGroup());
+        if (currentRecord.getReadPairedFlag()) {
+            if (read1 != null && read2 !=null){
+                writeTagRecords(read1, 1, rgTagWriters);
+                writeTagRecords(read2, 2, rgTagWriters);
+            }
+        } else {
+            writeTagRecords(currentRecord, null, rgTagWriters);
+        }
+    }
+
+    @Override
+    protected Map<SAMReadGroupRecord, List<FastqWriter>> generateAdditionalWriters(List<SAMReadGroupRecord> readGroups, FastqWriterFactory factory) {
+        return generateTagWriters(readGroups, factory);
+    }
+
+    // generate writers
+    private Map<SAMReadGroupRecord, List<FastqWriter>> generateTagWriters(final List<SAMReadGroupRecord> samReadGroupRecords,
+                                                                          final FastqWriterFactory factory) {
+        final Map<SAMReadGroupRecord, List<FastqWriter>> writerMap = new HashMap<>();
+
+        if (!OUTPUT_PER_RG) {
+            /* Prepare tag writers based on sequence tag groups provided in command line */
+
+            final List<FastqWriter> tagFastqWriters = makeTagWriters(null).stream().map(factory::newWriter).collect(Collectors.toList());
+
+            writerMap.put(null, tagFastqWriters);
+            for (final SAMReadGroupRecord rg : samReadGroupRecords) {
+                writerMap.put(rg, tagFastqWriters);
+            }
+        } else {
+            /* prepare tag writers based on readgroup names */
+            for (final SAMReadGroupRecord rg : samReadGroupRecords) {
+                List<FastqWriter> tagWriters = makeTagWriters(rg).stream().map(factory::newWriter).collect(Collectors.toList());
+
+                writerMap.put(rg, tagWriters);
+            }
+        }
+        return writerMap;
+    }
+
+    // Creates fastq writers based on readgroup passed in and sequence tag groupings from command line
+    private List<File> makeTagWriters(final SAMReadGroupRecord readGroup) {
```

Comment:
This is called twice and in both cases the result is passed to `.stream().map(factory::newWriter).collect(Collectors.toList())`. To avoid duplication that work could be done in this method, or another method that calls this method.

Or this method could be modified to call `factory.newWriter()` directly when before adding the new writer to the collection `tagFiles`.

Reply:
yea it would nice if the method returned what the comment/method name describe.....